src/docfx/lib/git/FileCommitProvider.cs

```diff
@@ -193,145 +192,162 @@ public void Dispose()
         {
-            if (git_object_lookup(out var tree, _repo, &treeId, 2 /* GIT_OBJ_TREE */) != 0)
+            var treeId = tree.Id;
+            if (git_object_lookup(out var pTree, _repo, &treeId, 2 /* GIT_OBJ_TREE */) != 0)
             {
-                return null;
+                return tree.Children = s_emptyTree;
             }
 
-            var n = git_tree_entrycount(tree);
-            var blobs = new Dictionary<int, git_oid>((int)n);
+            var n = (int)git_tree_entrycount(pTree);
+            var children = new DictionarySlim<uint, Tree>(n);
 
-            for (var p = IntPtr.Zero; p != n; p = p + 1)
+            for (var i = 0; i < n; i++)
             {
-                var entry = git_tree_entry_byindex(tree, p);
-                var name = Marshal.PtrToStringUTF8(git_tree_entry_name(entry)) ?? "";
+                var entry = git_tree_entry_byindex(pTree, (IntPtr)i);
+                var entryId = *git_tree_entry_id(entry);
+                var pName = git_tree_entry_name(entry);
+                var pNameEnd = pName;
+                while (*pNameEnd != 0)
+                {
+                    pNameEnd++;
+                }
 
-                blobs[GetStringId(name)] = *git_tree_entry_id(entry);
-            }
+                var name = new ReadOnlySpan<byte>(pName, (int)(pNameEnd - pName));
 
-            git_object_free(tree);
+                children.GetOrAddValueRef(HashUtility.GetFnv1A32Hash(name)) = _treeCache.GetOrAdd(entryId.a, (_, id) => new Tree { Id = id }, entryId);
+            }
 
-            return blobs;
-        }
+            git_object_free(pTree);
 
-        private int GetStringId(string value)
-        {
-            return _stringPool.GetOrAdd(value, _ => Interlocked.Increment(ref _nextStringId));
+            return tree.Children = children;
         }
 
         private class Commit
         {
-            public git_oid Sha { get; }
+            public git_oid Id { get; }
 
-            public git_oid Tree { get; }
+            public Tree Tree { get; }
 
             public GitCommit GitCommit { get; }
 
             public Commit[] Parents { get; }
 
-            public git_oid[] ParentShas { get; set; }
+            public git_oid[] ParentIds { get; set; }
 
-            public Commit(git_oid sha, git_oid[] parentShas, git_oid tree, GitCommit gitCommit)
+            public Commit(git_oid id, git_oid[] parentIds, Tree tree, GitCommit gitCommit)
             {
-                Sha = sha;
-                ParentShas = parentShas;
-                Parents = new Commit[parentShas.Length];
+                Id = id;
+                ParentIds = parentIds;
                 Tree = tree;
                 GitCommit = gitCommit;
+                Parents = new Commit[parentIds.Length];
             }
         }
+
+        private class Tree
+        {
+            public git_oid Id { get; set; }
+
+            public DictionarySlim<uint, Tree>? Children { get; set; }
```

Comment:
don't quite follow the reduction of dictionary lookup here... 
Seems like it needed 1 dictionary lookup before and 1 now